src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/BinaryArrayRecord.cs

```diff
@@ -0,0 +1,236 @@
+
+    private MemberTypeInfo MemberTypeInfo { get; }
+
+    private List<object> Values { get; }
+
+    [RequiresDynamicCode("May call Array.CreateInstance() and Type.MakeArrayType().")]
+    private protected override Array Deserialize(Type arrayType, bool allowNulls, int maxLength)
+    {
+        Type elementType = MapElementType(arrayType, out bool isClassRecord);
+        Type actualElementType = arrayType.GetElementType()!;
+        Array array =
+#if NET9_0_OR_GREATER
+            isClassRecord
+                ? Array.CreateInstance(elementType, Length)
+                : Array.CreateInstanceFromArrayType(arrayType, Length);
+#else
+            Array.CreateInstance(elementType, Length);
+#endif
+
+        int resultIndex = 0;
+        foreach (object value in Values)
+        {
+            object item = value is MemberReferenceRecord referenceRecord
+                ? referenceRecord.GetReferencedRecord()
+                : value;
+
+            if (item is not SerializationRecord record)
+            {
+                array.SetValue(item, resultIndex++);
+                continue;
+            }
+
+            switch (record.RecordType)
+            {
+                case RecordType.BinaryArray:
+                case RecordType.ArraySinglePrimitive:
+                case RecordType.ArraySingleObject:
+                case RecordType.ArraySingleString:
+                    ArrayRecord nestedArrayRecord = (ArrayRecord)record;
+                    Array nestedArray = nestedArrayRecord.ToArray(actualElementType, allowNulls, maxLength);
+                    array.SetValue(nestedArray, resultIndex++);
+                    break;
+                case RecordType.ObjectNull:
+                case RecordType.ObjectNullMultiple256:
+                case RecordType.ObjectNullMultiple:
+                    if (!allowNulls)
+                    {
+                        ThrowHelper.ThrowArrayContainedNulls();
+                    }
+
+                    int nullCount = ((NullsRecord)item).NullCount;
+                    do
+                    {
+                        array.SetValue(null, resultIndex++);
+                        nullCount--;
+                    }
+                    while (nullCount > 0);
+                    break;
+                default:
+                    array.SetValue(record.GetValue(), resultIndex++);
+                    break;
+            }
+        }
+
+        return array;
+    }
+
+    internal static ArrayRecord Parse(BinaryReader reader, RecordMap recordMap, PayloadOptions options)
+    {
+        int objectId = reader.ReadInt32();
+        BinaryArrayType arrayType = reader.ReadArrayType();
+        int rank = reader.ReadInt32();
+
+        bool isRectangular = arrayType is BinaryArrayType.Rectangular or BinaryArrayType.RectangularOffset;
+
+        if (rank < 1 || rank > 32
+            || (rank != 1 && !isRectangular)
+            || (rank == 1 && isRectangular))
```

Comment:
Why is the reader being opinionated that you can't use BinaryArrayRecord for a szArray?  I don't see anything in 2.4.3.1 that supports that block.

I also don't know where the rank-32 limit is coming from; or why you can't read `int[][,]` (a Jagged array with rank 2 on the outer arrayness).

This needs a lot of comments.

Reply:
> Why is the reader being opinionated that you can't use BinaryArrayRecord for a szArray?

It allows for SZArrays, actually `BinaryArrayRecord` is the only thing that can represent a SZArray of classes/structs other that `system.object`.

> I also don't know where the rank-32 limit is coming from

This is a C# limit (you can't create arrays with rank larger than 32)

> why you can't read int[][,] (a Jagged array with rank 2 on the outer arrayness).

I am going to add a test for that

Reply:
> This is a C# limit (you can't create arrays with rank larger than 32)

This is not C# limit. C# will compile arrays with ranks larger than 32 just fine. It is an arbitrary limit in the current CoreCLR type loader.